BUG: GH11206 where pd.isnull did not consider numpy NaT null

kawochen opened the pull request:
closes #11206

kawochen committed:
BUG: GH11206 where pd.isnull did not consider numpy NaT null

jreback commented:
lgtm.

can you run a perf check on the null benchmarks. just confirm nothing has substantially changed.

kawochen commented:
Updated. Timing for `frame_methods.frame_isnull.time_frame_isnull` (where all data is non-`nan` `float`s) increased by less than 5%.
Added a benchmark where lots of the data is various types of null.
```
BENCHMARK                                    BEFORE      AFTER     FACTOR
...ame_isnull_has_null.time_frame_isnull    38.18ms   138.88ms   3.63779915x
```

I also changed the code slightly so that the logic is further short-circuited short-circuited.

jreback commented:
the benchmark u are showing is 3x increase?

kawochen commented:
Yes.That's the with lots of nulls. I will dig a little to see if that can be avoided.

jreback commented:
these routines are hit *everywhere*, so i am for special casing certain things. IOW, that is why ``isnullobj`` exists. But it may make sense to have 2 versions of ``_checknull_with_nat``, because they are called on different array kinds. E.g if you have an int array then it obviously doesn't need to check for ``None,NaN,NaT`` at all. most of these cases are already handled, but IIRC, this is hotspotted. So just want to try to nail that down.

kawochen commented:
Updated.

kawochen commented on pandas/lib.pyx:
```diff
@@ -321,11 +320,11 @@ def isnullobj(ndarray[object] arr):
-    return result.view(np.bool_)
+        result[i] = _check_all_nulls(val)
+    return result.view(dtype=np.bool_)
 
```
I forgot to remove this.  I will remove in the next update.

jreback replied:
what are you fixing here?

kawochen replied:
you mean the `_check_all_nulls`?  Only object arrays hit this path so I thought it'd be safe to have the most generic version of `checknull` here?

For the `dtype` thing, I forgot to remove it after some experiment, but will remove it in the next update.

jreback replied:
yep that is fine

jreback commented on asv_bench/benchmarks/frame_methods.py:
```diff
@@ -582,7 +582,7 @@ def time_frame_interpolate_some_good_infer(self):
 
 
-class frame_isnull(object):
+class frame_isnull_flaots_no_null(object):
```
``floats`` typo

jreback commented on pandas/tslib.pyx:
```diff
@@ -707,12 +708,26 @@ NaT = NaTType()
+cdef inline bint _check_all_nulls(object val):
+    """ utility to check if a value is any type of null """
+    cdef bint res = 0
+    if PyFloat_Check(val):
```
shouldn't you just return at each stage? rather than setting a value then returning it at the end

kawochen commented:
udpated

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -1146,7 +1146,7 @@ Bug Fixes
 - Bug in indexing with a ``PeriodIndex`` on an object with a ``PeriodIndex`` (:issue:`4125`)
 - Bug in ``read_csv`` with ``engine='c'``: EOF preceded by a comment, blank line, etc. was not handled correctly (:issue:`10728`, :issue:`10548`)
-
+- Bug in ``isnull`` where ``numpy.datetime64('NaT')`` in a ``numpy.array`` was not determined to be null(:issue:`11206`)
```
let's put this in 0.17.1

jreback replied:
pls move to 0.17.1

kawochen commented:
ok. I'm traveling and will get to this next week.

jreback commented:
@kawochen can you update (move whatsnew to 0.17.1)

kawochen commented:
Will do (still on vacation)

> On Oct 9, 2015, at 8:35 AM, Jeff Reback <notifications@github.com> wrote:
> 
> @kawochen can you update (move whatsnew to 0.17.1)
> 
> —
> Reply to this email directly or view it on GitHub.
>

kawochen commented:
updated

jreback commented:
can u rebase and run a perf check

kawochen commented:
Saw some horrifying perf hit (30X).  I will look at the compiled code later.

jreback commented:
ok gr8!

kawochen commented:
so using `timeit` and this code
```
from pandas import *
import pandas as pd
import numpy as np

np.random.seed(1234)
sample = np.array([NaT, np.nan, None, np.datetime64('NaT'),
    np.timedelta64('NaT'), 0, 1, 2.0, '', 'abcd'])
data = np.random.choice(sample, (1000, 1000))
df = DataFrame(data)

%timeit isnull(df)
```
the perf hit is 4X.  But if I that put into an `asv` benchmark, it shows 24X. I seem to be having problems with `asv` lately -- not sure what's wrong.

jreback commented:
are you using ``profile=True`` in the cython modules?

pv commented:
The perf hit is different for numpy 1.10 vs 1.9

jreback commented:
@kawochen pushing, but if you finish in the next few days, pls ping

kawochen commented:
@pv thanks! :smile: 
@jreback 
branch
`
[ 32.58%] ··· Running ...s.frame_isnull_obj.time_frame_isnull    46.07ms
`
master
`
[ 82.58%] ··· Running ...s.frame_isnull_obj.time_frame_isnull    38.91ms
`

`

jreback commented on pandas/tslib.pyx:
```diff
@@ -707,12 +708,28 @@ NaT = NaTType()
 
 iNaT = util.get_nat()
 
-
```
maybe blow away ``_checknull_with_nat``(and replace with ``_check_all_nulls``  and see if any timings are affected?

kawochen commented:
Replacing wouldn't work in all cases, because sometimes you don't want `timedelta64` to be accepted, e.g. `datetime_to_datetime64()`, so it would become some sort of manual inlining for `_checknull_with_nat` to deal with those cases (which might not be a bad thing, depending on how often this is needed).

But I see that `tslib.pyx` needs some cleaning, as code like `if val is np_NaT or val.view('i8') == iNaT` is everywhere (`np_NaT` is not a singleton, so the first half of the expression is almost certainly `False`, and the `val.view('i8')` is slower than `get_datetime64_val` (especially given the regression in numpy 1.10, but that will be fixed soon).

So maybe when others or I clean up this module, we can see how to organize or deal with all the lower level scalar check null functions

jreback commented:
@kawochen thanks!

always nice PR's from you! keep it up!
